torch.Package zipfile debugging printer

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52176 torch.Package zipfile debugging printer**
* #52218 torch.Package extend PyTorchStreamWriter to track written records

Added tooling to print out zipfile structure for PackageExporter and PackageImporter.

API looks like: 
```
importer.file_structure() // str representation prints all files 
importer.file_structure(exclude="**/*.storage") // str representation prints all files, excluding storage files
e.file_structure(include="**/utils.py") // str representation only includes file structure on utilt.py's path
```

The `importer.file_structure()` str representation of a packaged resent look like this: 
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── 94304869784064.storage
    │   ├── 94304870911616.storage
    │   ├── 94304900784016.storage
    │   ├── 94304901821024.storage
         ** many many more .storages **
    │   ├── 94304908412240.storage
    │   ├── 94304908414512.storage
    │   ├── 94304908414944.storage
    │   ├── 94304908416608.storage
    │   ├── 94304908421408.storage
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── model_1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
`importer.file_structure(exclude="**/*.storage")`:
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── model_1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
`e.file_structure(include="**/utils.py")`:
```
─── resnet_added_attr_test
    └── torchvision
        └── models
            └── utils.py
```
Differential Revision: [D26429795](https://our.internmc.facebook.com/intern/diff/D26429795)

Commit:
torch.Package zipfile debugging printer

[ghstack-poisoned]

Lilyjjo commented on test/test_package.py:
```diff
@@ -82,6 +82,32 @@ def test_save_module(self):
+            he.print_file_structure("init")
+        f.seek(0)
+        hi = PackageImporter(f) 
+        hi.print_file_structure()
```
This test might be annoying to have in the test suite because it'll print out the file structure when the test is ran

Lilyjjo replied:
Now test doesn't check output, only makes sure API works

Lilyjjo commented on torch/package/_package_utils.py:
```diff
@@ -0,0 +1,60 @@
+            top_folder.add_file(file)
+    str_list: List[str] = []
+    top_folder.stringify_tree(str_list)
+    print("".join(str_list))
```
I wasn't sure if either printing out or returning the file structure as a string would be a better user experience. This PR goes with printing out the file structure, but I can change it

zdevito replied:
Probably return the string. It's easy to print a string, it is hard to make a string out of something printed to stdout.

Lilyjjo replied:
Cool, changed to the string

zdevito commented on torch/package/_package_utils.py:
```diff
@@ -0,0 +1,60 @@
+import re
```
`_package_utils.py` -> `_file_structure_printer.py` 

It's not a good idea to name file utils. In the limit you end up with things like `utils/utils.py` with every function defined in it (I've literally just be working with a repo like this).

zdevito commented on torch/package/exporter.py:
```diff
@@ -69,12 +70,15 @@ def __init__(self, f: Union[str, Path, BinaryIO], verbose: bool = True):
 
         self.zip_file = torch._C.PyTorchFileWriter(f)
         self.zip_file.set_min_version(6)
+        self._written_records : List[str] = []
```
This approach duplicates the list of records here and in the zip file. It's not a great idea since eventually this will share a zip archive with the JIT which won't use the python API for putting the stuff in the archive. So this would miss those files.

zdevito replied:
self.name is similar: the zip file knows it already, so duplicating it would make it harder.

zdevito replied:
It should be possible for the zip file object to track the files it has written so that it can also implement `get_all_records` similar to the reader. The zip file already has an archiveName function, maybe just needs to be bound into Python.

Lilyjjo replied:
I moved the generation of the records list functionality to the PythonStreamWriter in CPP, it's now in the PR below this one 👍

zdevito commented on torch/package/_package_utils.py:
```diff
@@ -0,0 +1,60 @@
+
+
+def _print_file_structure(
+    filename: str, records: List[str], skip_storage: bool = True, path_includes=".*"
```
it feels odd to have skip_storage be a very specific way to exclude storages, but path_includes be a very generic way to include files. Also in the package importer are mock and extern functions which take exclude/include pairs using file glob syntax and not regex, so using regex here feels inconsistent with that. Maybe use `glob` for both.

zdevito reviewed (commented):
Cool! This will be really useful figuring out how to build shared zip archives. I have some minor comments mostly around future proofing for when we will use this with JIT as well, and making it consistent with other parts of the packaging API.

Commit:
Update on "torch.Package zipfile debugging printer"


Added tooling to print out zipfile structure for PackageExporter and PackageImporter.

API looks like: 
```
exporter.print_file_structure("sss" /*only include files with this in the path*/)
importer3.print_file_structure(False /*don't print storage*/, "sss" /*only include files with this in the path*/)
```


The output looks like this with the storage hidden by default:
```
─── resnet.zip
    ├── .data
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── models1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
The output looks like this with the storage being printed out: 
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── 94574437434544.storage
    │   ├── 94574468343696.storage
    │   ├── 94574470147744.storage
    │   ├── 94574470198784.storage
    │   ├── 94574470267968.storage
    │   ├── 94574474917984.storage
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── models1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```

If the output is filtered with the string 'utils' it'd looks like this:
```
─── resnet_added_attr_test.zip
    └── torchvision
        └── models
            └── utils.py
```


[ghstack-poisoned]

Lilyjjo commented on torch/package/_file_structure_printer.py:
```diff
@@ -0,0 +1,61 @@
+
+    top_folder = Folder(filename)
+    for file in records:
+        if pathlib.PurePath(file).match(include) and (
```
I can probably refactor the code to use the _GlobGroup from exporter.py if prefered, but this seemed easier

Commit:
Update on "torch.Package zipfile debugging printer"


Added tooling to print out zipfile structure for PackageExporter and PackageImporter.

API looks like: 
```
exporter.print_file_structure("sss" /*only include files with this in the path*/)
importer3.print_file_structure(False /*don't print storage*/, "sss" /*only include files with this in the path*/)
```


The output looks like this with the storage hidden by default:
```
─── resnet.zip
    ├── .data
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── models1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
The output looks like this with the storage being printed out: 
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── 94574437434544.storage
    │   ├── 94574468343696.storage
    │   ├── 94574470147744.storage
    │   ├── 94574470198784.storage
    │   ├── 94574470267968.storage
    │   ├── 94574474917984.storage
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── models1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```

If the output is filtered with the string 'utils' it'd looks like this:
```
─── resnet_added_attr_test.zip
    └── torchvision
        └── models
            └── utils.py
```


[ghstack-poisoned]

Commit:
Update on "torch.Package zipfile debugging printer"


Added tooling to print out zipfile structure for PackageExporter and PackageImporter.

API looks like: 
```
exporter.print_file_structure("sss" /*only include files with this in the path*/)
importer3.print_file_structure(False /*don't print storage*/, "sss" /*only include files with this in the path*/)
```


The output looks like this with the storage hidden by default:
```
─── resnet.zip
    ├── .data
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── models1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
The output looks like this with the storage being printed out: 
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── 94574437434544.storage
    │   ├── 94574468343696.storage
    │   ├── 94574470147744.storage
    │   ├── 94574470198784.storage
    │   ├── 94574470267968.storage
    │   ├── 94574474917984.storage
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── models1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```

If the output is filtered with the string 'utils' it'd looks like this:
```
─── resnet_added_attr_test.zip
    └── torchvision
        └── models
            └── utils.py
```

Differential Revision: [D26429795](https://our.internmc.facebook.com/intern/diff/D26429795)

[ghstack-poisoned]

suo commented on torch/package/_file_structure_printer.py:
```diff
@@ -0,0 +1,61 @@
+    class Folder:
+        def __init__(self, name: str):
+            self.name = name
+            self.sub_folders: Dict[str, Folder] = {}
```
do we actually need to track folders and files are separately? afaict, a folder with no elements looks exactly the same as a file in the tree right? So potentially we could have just `self.children` and `self.is_dir`, which follows the unix way of doing it a little more closely.

suo reviewed (changes requested):
As a general rule, it's good to return a structured object from an API instead of a string. For example, if users wanted to use the structure in any way besides printing, they'd be forced to parse the string representation or process the zipfile itself.
- Let's make `Folder` a public class and make the construction methods internal.
- Make a `__str__` method of `Folder` do the tree stringification.
 
 That way `print(exporter.print_file_structure())` behaves identically to before, but users get more flexibility with the results

Commit:
Update on "torch.Package zipfile debugging printer"


Added tooling to print out zipfile structure for PackageExporter and PackageImporter.

API looks like: 
```
exporter.print_file_structure() // prints all files 
exporter.print_file_structure("sss") // only print files with "sss" in path
importer.print_file_structure("sss", "**/*.py" ) // only print files with "sss" in path and exclude any files ending with .py
```

The output looks like this: 
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── 94574437434544.storage
    │   ├── 94574468343696.storage
    │   ├── 94574470147744.storage
    │   ├── 94574470198784.storage
    │   ├── 94574470267968.storage
    │   ├── 94574474917984.storage
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── models1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```

If the output is filtered with the string 'utils.py' it'd looks like this:

```
─── resnet_added_attr_test.zip
    └── torchvision
        └── models
            └── utils.py
```

Differential Revision: [D26429795](https://our.internmc.facebook.com/intern/diff/D26429795)

[ghstack-poisoned]

Lilyjjo commented on torch/package/_glob_group.py:
```diff
@@ -0,0 +1,44 @@
+        return self._dbg
+
+    def matches(self, candidate: str) -> bool:
+        if candidate[:5] != ".data":
```
The only change I made to the _GlobGroup was preventing the `.data` from having the `.` prepended as it didn't allow the `.data` folder to be matched to

zdevito replied:
This glob infrastructure is for package names, and assumes that things are separated by periods. But files are separated by `/` and might contain periods. I feel like this should just take the separate token as an argument to the GlobalGroup, and then this special case would not need to exist.

Lilyjjo replied:
Cool, I modified the GlobGroup to accept whatever separators. I can change the GlobGroup's `_glob_to_re` and `_glob_list` to not be staticmethods if that makes the code cleaner (I wasn't sure if they were staticmethods for a reason)

Commit:
Update on "torch.Package zipfile debugging printer"


Added tooling to print out zipfile structure for PackageExporter and PackageImporter.

API looks like: 
```
importer.file_structure() // str representation prints all files 
importer.file_structure(exclude="*.storage") // str representation prints all files, excluding storage files
e.file_structure(include="*utils.py") // str representation only includes file structure on utilt.py's path
```

The `importer.file_structure()` str representation of a packaged resent look like this: 
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── 94304869784064.storage
    │   ├── 94304870911616.storage
    │   ├── 94304900784016.storage
    │   ├── 94304901821024.storage
         ** many many more .storages **
    │   ├── 94304908412240.storage
    │   ├── 94304908414512.storage
    │   ├── 94304908414944.storage
    │   ├── 94304908416608.storage
    │   ├── 94304908421408.storage
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── model_1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
`importer.file_structure(exclude="*.storage")`:
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── model_1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
`e.file_structure(include="*utils.py")`:
```
─── resnet_added_attr_test
    └── torchvision
        └── models
            └── utils.py
```
Differential Revision: [D26429795](https://our.internmc.facebook.com/intern/diff/D26429795)

[ghstack-poisoned]

suo commented on torch/package/exporter.py:
```diff
@@ -125,6 +126,19 @@ def save_source_file(self, module_name: str, file_or_directory: str, dependencie
             is_package = path.name == '__init__.py'
             self.save_source_string(module_name, _read_file(file_or_directory), is_package, dependencies, file_or_directory)
 
+    def file_structure(self, *, include: 'GlobPattern' = "**", exclude: 'GlobPattern' = ()) -> str:
```
this no longer returns a str

suo reviewed (approved):
lgtm! minor comment inline

Commit:
Update on "torch.Package zipfile debugging printer"


Added tooling to print out zipfile structure for PackageExporter and PackageImporter.

API looks like: 
```
importer.file_structure() // str representation prints all files 
importer.file_structure(exclude="**/*.storage") // str representation prints all files, excluding storage files
e.file_structure(include="**/utils.py") // str representation only includes file structure on utilt.py's path
```

The `importer.file_structure()` str representation of a packaged resent look like this: 
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── 94304869784064.storage
    │   ├── 94304870911616.storage
    │   ├── 94304900784016.storage
    │   ├── 94304901821024.storage
         ** many many more .storages **
    │   ├── 94304908412240.storage
    │   ├── 94304908414512.storage
    │   ├── 94304908414944.storage
    │   ├── 94304908416608.storage
    │   ├── 94304908421408.storage
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── model_1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
`importer.file_structure(exclude="**/*.storage")`:
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── model_1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
`e.file_structure(include="**/utils.py")`:
```
─── resnet_added_attr_test
    └── torchvision
        └── models
            └── utils.py
```
Differential Revision: [D26429795](https://our.internmc.facebook.com/intern/diff/D26429795)

[ghstack-poisoned]

Commit:
Update on "torch.Package zipfile debugging printer"


Added tooling to print out zipfile structure for PackageExporter and PackageImporter.

API looks like: 
```
importer.file_structure() // str representation prints all files 
importer.file_structure(exclude="**/*.storage") // str representation prints all files, excluding storage files
e.file_structure(include="**/utils.py") // str representation only includes file structure on utilt.py's path
```

The `importer.file_structure()` str representation of a packaged resent look like this: 
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── 94304869784064.storage
    │   ├── 94304870911616.storage
    │   ├── 94304900784016.storage
    │   ├── 94304901821024.storage
         ** many many more .storages **
    │   ├── 94304908412240.storage
    │   ├── 94304908414512.storage
    │   ├── 94304908414944.storage
    │   ├── 94304908416608.storage
    │   ├── 94304908421408.storage
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── model_1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
`importer.file_structure(exclude="**/*.storage")`:
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── model_1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
`e.file_structure(include="**/utils.py")`:
```
─── resnet_added_attr_test
    └── torchvision
        └── models
            └── utils.py
```
Differential Revision: [D26429795](https://our.internmc.facebook.com/intern/diff/D26429795)

[ghstack-poisoned]

Commit:
Update on "torch.Package zipfile debugging printer"


Added tooling to print out zipfile structure for PackageExporter and PackageImporter.

API looks like: 
```
importer.file_structure() // str representation prints all files 
importer.file_structure(exclude="**/*.storage") // str representation prints all files, excluding storage files
e.file_structure(include="**/utils.py") // str representation only includes file structure on utilt.py's path
```

The `importer.file_structure()` str representation of a packaged resent look like this: 
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── 94304869784064.storage
    │   ├── 94304870911616.storage
    │   ├── 94304900784016.storage
    │   ├── 94304901821024.storage
         ** many many more .storages **
    │   ├── 94304908412240.storage
    │   ├── 94304908414512.storage
    │   ├── 94304908414944.storage
    │   ├── 94304908416608.storage
    │   ├── 94304908421408.storage
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── model_1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
`importer.file_structure(exclude="**/*.storage")`:
```
─── resnet_added_attr_test.zip
    ├── .data
    │   ├── extern_modules
    │   └── version
    ├── models
    │   └── model_1.pkl
    └── torchvision
        └── models
            ├── resnet.py
            └── utils.py
```
`e.file_structure(include="**/utils.py")`:
```
─── resnet_added_attr_test
    └── torchvision
        └── models
            └── utils.py
```
Differential Revision: [D26429795](https://our.internmc.facebook.com/intern/diff/D26429795)

[ghstack-poisoned]
